### DISCUSSION TITLE
Ways around the Clone requirement on Message?

### DISCUSSION THREAD

**Discussion by anderslanglands** [hecrj/iced#1208] [Q&A] [answered]
Using certain widgets (e.g. Button) adds a Clone requirement on Message. As far as I can tell you can only have one Message type for your application, and Messages seem to be the way to pass async data around. This combination can be a little restrictive. Is there some way of structuring my Messages to avoid the clone requirement? Or should I just be sticking everything inside an Arc and not worrying about it?

**Comment by MoAlyousef** [comment]
Removing the Clone bound will end up being more restrictive, since it means Copy types will be copied and non-copyable types will be moved.

**Comment by Eucladia** [comment, marked as answer]
You can decouple the required variants that need to be `Clone` in your current `Message` into another enum that is `Clone`, use that `Message` for your widgets, convert the `Widget` into an `Element` with that cloneable message, then use `Element::map` to convert the cloneable `Message` into the non-clone `Message`. 

You can check [this](https://github.com/hecrj/iced/pull/155#issuecomment-573523405) out to see what the code for that would look like.

**Comment by anderslanglands** [comment]
Thanks for that! I was looking for something like this but hadn’t found
Element::map.
